### PR TITLE
ExprAmount Recursive Size of List

### PR DESCRIPTION
**Pull request by Wealthyturtle**
### Description
Adds another pattern to ExprAmount that allows users to get the recursive size of lists (only applies to list variables)

---
**Target Minecraft Versions:**     Any
**Requirements:**     None
**Related Issues:**     #1869

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprAmount.java [hunk 1]
```diff
@@ -41,14 +45,20 @@
  */
 @Name("Amount")
 @Description({"The amount of something.",
-		"Please note that <code>amount of %items%</code> will not return the number of items, but the number of stacks, e.g. 1 for a stack of 64 torches. To get the amount of items in a stack, see the <a href='#ExprItemAmount'>item amount</a> expression."})
+		"Please note that <code>amount of %items%</code> will not return the number of items, but the number of stacks, e.g. 1 for a stack of 64 torches. To get the amount of items in a stack, see the <a href='#ExprItemAmount'>item amount</a> expression.",
+		"Also please note that getting a list's recursive size can cause lag if the list is large."})
```
Comment (bensku) [inline src/main/java/ch/njol/skript/expressions/ExprAmount.java]: Documentation for recursive size is insufficient. If even I don't know what exactly it does without reading implementation, something is wrong.

File: src/main/java/ch/njol/skript/expressions/ExprAmount.java [hunk 2]
```diff
@@ -77,12 +105,32 @@ public boolean isSingle() {
...
+			if (var != null)
+				return new Integer[] {getRecursiveSize((TreeMap<String, Object>) var)}; // Should already be a TreeMap 
+		}
 		return new Integer[] {expr.getArray(e).length};
 	}
 	
+	@SuppressWarnings("unchecked")
+	private static int getRecursiveSize(TreeMap<String, Object> map) {
+		int count = 0;
+		for (Map.Entry<String, Object> entry : map.entrySet()) {
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/expressions/ExprAmount.java]: since you don't care about the key, you can just use map.values()

Comment (Pikachu920) [inline src/main/java/ch/njol/skript/expressions/ExprAmount.java]: `*` index is actually a value, do you want to be counting that?

File: src/main/java/ch/njol/skript/expressions/ExprAmount.java [hunk 3]
```diff
@@ -77,12 +105,32 @@ public boolean isSingle() {
...
+			Object value = entry.getValue();
+			count += (value instanceof TreeMap) ? getRecursiveSize((TreeMap<String, Object>) value) : 1;
```
[the remark this answers is not in the document]
Comment (Wealthyturtle) [reply src/main/java/ch/njol/skript/expressions/ExprAmount.java]: Is this for neatness?

**Comment by Wealthyturtle** [comment]
Someone requested this feature so I thought I'd implement it anyways, could be useful in certain instances.

File: src/main/java/ch/njol/skript/expressions/ExprAmount.java [hunk 4]
```diff
@@ -41,14 +45,30 @@
...
 @Since("1.0")
 public class ExprAmount extends SimpleExpression<Integer> {
 	static {
-		Skript.registerExpression(ExprAmount.class, Integer.class, ExpressionType.PROPERTY, "(amount|number|size) of %objects%");
+		Skript.registerExpression(ExprAmount.class, Integer.class, ExpressionType.PROPERTY, 
+				"(amount|number|size) of %objects%",
+				"(recursive|total) (amount|number|size) of %objects%");
 	}
 	
+	private String exprString = "";
```
[the remark this answers is not in the document]
Comment (Wealthyturtle) [reply src/main/java/ch/njol/skript/expressions/ExprAmount.java]: Not sure why I did that initially, thanks for catching that!

File: src/main/java/ch/njol/skript/expressions/ExprAmount.java (hunk 2 above)
Comment (Wealthyturtle) [reply src/main/java/ch/njol/skript/expressions/ExprAmount.java]: Sorry, I'm not too sure what you mean.
I added a sysout (``System.out.println(entry.getKey() + "," + entry.getValue().getClass().getCanonicalName());``), and got this as output:
```
1,java.lang.String
2,java.util.TreeMap
null,java.lang.String
1,java.lang.String
2,java.lang.String
4
```
For the script:
```
command /test:
	trigger:
		set {_t::1} to "a"
		set {_t::2} to "b"
		set {_t::2::*} to "c"
		set {_t::2::1} to "d"
		set {_t::2::2} to "e"
		broadcast "%recursive size of {_t::*}%"
```

File: src/main/java/ch/njol/skript/expressions/ExprAmount.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Pikachu920) [reply src/main/java/ch/njol/skript/expressions/ExprAmount.java]: yeah basically

File: src/main/java/ch/njol/skript/expressions/ExprAmount.java (hunk 2 above)
Comment (Pikachu920) [reply src/main/java/ch/njol/skript/expressions/ExprAmount.java]: nevermind, then. must have been some mundosk specific weirdness

**Pull request merged**